### PR TITLE
CDRIVER-5648 fix libfuzzer integration

### PR DESCRIPTION
**Pull request by spencerjackson**
# Summary
This patch modifies the CMake rules in order to allow compilation of the already existing libfuzzer entrypoint.

# What Changed?
1. fuzz_test_libbson.c  renamed to fuzz_test_init_from_json.c
We could have multiple entrypoints, so it makes sense to name the file after the function we're fuzzing.
2. fuzz_test_init_from_json.c simplified to remove memory allocation
The faster we can run the entrypoint, the more iterations per second we'll be able to manage.
3. CMake option `MONGO_FUZZ` added
When building a fuzzer, *all* code in all TUs must be compiled with coverage instrumentation. The final link step must statically link libfuzzer's `main` symbol. Adding sanitizers to all TUs is also beneficial.
This new option enables AUBSan and a sanitizer called "fuzzer-no-link", which enables CovSAN and all the other flags libfuzzer needs to operate. Finally, this option sets `MONGO_FUZZ_ENGINE` to the linker flags needed to statically link in libfuzzer's `main`. Any binaries which provide fuzzer entrypoints need to use `MONGO_FUZZ_ENGINE` in their `LINK_FLAGS`.
5. Added `src/libbson/fuzz/CMakeLists.txt`
This file compiles the fuzzer entrypoint, iff `MONGO_FUZZ` is enabled.

I haven't used CMake in a while, so feedback is appreciated!

### CODE REVIEW

File: build/cmake/Fuzzer.cmake [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+mongo_setting(
+  MONGO_FUZZ "Whether fuzz testing is enabled"
+  TYPE BOOL
+  DEFAULT FALSE
+)
+
+if (MONGO_FUZZ)
+set(CMAKE_C_FLAGS "-fsanitize=undefined -fsanitize=address -fsanitize=fuzzer-no-link" CACHE STRING "Custom comp" FORCE)
+set(MONGO_FUZZ_ENGINE -fsanitize=fuzzer)
+endif ()
```
Comment (eramongodb) [inline build/cmake/Fuzzer.cmake:14]:
```suggestion
```
Remove in favor of reusing the existing `MONGO_SANITIZE` option. Add the following after the `MONGO_SANITIZE` option definition in `build/cmake/Sanitizers.cmake`:

```cmake
if (MONGO_FUZZ)
    set(mongo_fuzz_options "address,undefined,fuzzer-no-link")

    if (NOT "${MONGO_SANITIZE}" STREQUAL "${mongo_fuzz_options}")
        message(WARNING "Overriding user-provided MONGO_SANITIZE options in favor of MONGO_SANITIZE=ON")
    endif()

    set_property (CACHE MONGO_SANITIZE PROPERTY VALUE "${mongo_fuzz_options}")
endif ()
```

File: src/libbson/fuzz/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+if (MONGO_FUZZ)
+
+add_executable(fuzz_test_init_from_json
+   fuzz_test_init_from_json.c)
```
Comment (eramongodb) [inline src/libbson/fuzz/CMakeLists.txt:4]:
```suggestion
add_executable(fuzz_test_init_from_json EXCLUDE_FROM_ALL fuzz_test_init_from_json.c)
```
Do not build fuzz executable as part of build-and-install routines.

File: src/libbson/fuzz/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+target_include_directories(fuzz_test_init_from_json PRIVATE "src" "${PROJECT_BINARY_DIR}/src/")
+target_link_libraries(fuzz_test_init_from_json ${bson_libs})
+set_target_properties(fuzz_test_init_from_json PROPERTIES LINK_FLAGS ${MONGO_FUZZ_ENGINE})
```
Comment (eramongodb) [inline src/libbson/fuzz/CMakeLists.txt:7]:
```suggestion
set_property(TARGET fuzz_test_init_from_json APPEND PROPERTY LINK_OPTIONS -fsanitize=fuzzer)
```
Pass fuzzer "with link" option directly and avoid potentially overwriting existing `LINK_OPTIONS` values.

**Review by eramongodb — CHANGES_REQUESTED** [review]

File: build/cmake/Fuzzer.cmake
[the remark this answers is not in the document]
Comment (spencerjackson) [reply build/cmake/Fuzzer.cmake:1]: Done

File: build/cmake/Fuzzer.cmake [hunk 4]
```diff
@@ -0,0 +1,14 @@
+include (CheckCSourceCompiles)
+include (CMakePushCheckState)
+include (MongoSettings)
+
+mongo_setting(
+  MONGO_FUZZ "Whether fuzz testing is enabled"
+  TYPE BOOL
+  DEFAULT FALSE
+)
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply build/cmake/Fuzzer.cmake]: Done

File: build/cmake/Fuzzer.cmake (hunk 1 above)
Comment (spencerjackson) [reply build/cmake/Fuzzer.cmake]: Done

File: src/libbson/CMakeLists.txt [hunk 5]
```diff
@@ -399,6 +399,8 @@ install (EXPORT bson-targets
 include (LegacyPackage)
 include (CPack)
 
+add_subdirectory(fuzz)
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply src/libbson/CMakeLists.txt]: Done

File: src/libbson/fuzz/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+if (MONGO_FUZZ)
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply src/libbson/fuzz/CMakeLists.txt]: Done

File: src/libbson/fuzz/CMakeLists.txt (hunk 2 above)
Comment (spencerjackson) [reply src/libbson/fuzz/CMakeLists.txt]: Done

File: src/libbson/fuzz/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+add_executable(fuzz_test_init_from_json
+   fuzz_test_init_from_json.c)
+target_include_directories(fuzz_test_init_from_json PRIVATE "src" "${PROJECT_BINARY_DIR}/src/")
+target_link_libraries(fuzz_test_init_from_json ${bson_libs})
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply src/libbson/fuzz/CMakeLists.txt]: Done

File: src/libbson/fuzz/CMakeLists.txt (hunk 3 above)
Comment (spencerjackson) [reply src/libbson/fuzz/CMakeLists.txt]: Done

**Review by eramongodb — APPROVED** [review]
Thank you very much! One last suggestion remaining; otherwise, LGTM.

File: build/cmake/Sanitizers.cmake [hunk 8]
```diff
@@ -40,3 +40,21 @@ if (_sanitize)
...
+    ]]
+)
+
+if (MONGO_FUZZ)
+    set(mongo_fuzz_options "address,undefined,fuzzer-no-link")
+    if (NOT "${MONGO_SANITIZE}" STREQUAL "${mongo_fuzz_options}")
+        message(WARNING "Overriding user-provided MONGO_SANITIZE options in favor of MONGO_SANITIZE=ON")
+    endif ()
+    set_property (CACHE MONGO_SANITIZE PROPERTY VALUE "${mongo_fuzz_options}")
+endif ()
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply build/cmake/Sanitizers.cmake]: Done

**Comment by spencerjackson** [comment]
Comments addressed, I also tweaked the logic in the MONGO_FUZZ VALIDATE block, so that it errors only `if (MONGO_FUZZ AND NOT ENABLE_STATIC)`

**Review by kevinAlbs — APPROVED** [review]
Thank you for the contribution. LGTM with minor CMake suggested tweaks.

File: build/cmake/Sanitizers.cmake [hunk 9]
```diff
@@ -11,6 +11,24 @@ mongo_setting (
...
+        if (MONGO_FUZZ AND NOT ENABLE_STATIC)
+	    message (FATAL_ERROR "MONGO_FUZZ requires ENABLE_STATIC=ON or ENABLE_STATIC=BUILD_ONLY")
+	endif ()
+    ]]
+)
+
+if (MONGO_FUZZ)
+    set(mongo_fuzz_options "address,undefined,fuzzer-no-link")
+    if (NOT "${MONGO_SANITIZE}" STREQUAL "${mongo_fuzz_options}")
+        message(WARNING "Overriding user-provided MONGO_SANITIZE options in favor of MONGO_SANITIZE=ON")
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply build/cmake/Sanitizers.cmake]: Done

File: build/cmake/Sanitizers.cmake [hunk 10]
```diff
@@ -11,6 +11,24 @@ mongo_setting (
...
+    VALIDATE CODE [[
+        if (MONGO_FUZZ AND NOT ENABLE_STATIC)
+	    message (FATAL_ERROR "MONGO_FUZZ requires ENABLE_STATIC=ON or ENABLE_STATIC=BUILD_ONLY")
+	endif ()
+    ]]
+)
+
+if (MONGO_FUZZ)
+    set(mongo_fuzz_options "address,undefined,fuzzer-no-link")
+    if (NOT "${MONGO_SANITIZE}" STREQUAL "${mongo_fuzz_options}")
```
[the remark this answers is not in the document]
Comment (spencerjackson) [reply build/cmake/Sanitizers.cmake]: Done

**Pull request merged**